src/Castle.ts

```diff
@@ -0,0 +1,218 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+import AbortController from 'abort-controller';
+
+const defaultApiUrl = 'https://api.castle.io';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits: any;
+  created_at?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type AuthenticateResult = {
+  action: string;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+const isTimeout = (e: Error) => e.name === 'AbortError';
+
+export class Castle {
+  private apiSecret: string;
+  private apiUrl: string;
+  private timeout: number;
+  private allowedHeaders: string[];
+  private disallowedHeaders: string[];
+  private overrideFetch: any;
+  private failoverStrategy: string;
+
+  constructor({
+    apiSecret,
+    apiUrl,
+    timeout = 500,
+    allowedHeaders = [],
+    disallowedHeaders = [],
+    overrideFetch = fetch,
+    failoverStrategy = 'allow',
+  }: {
+    apiSecret: string;
+    apiUrl?: string;
+    timeout?: number;
+    allowedHeaders?: string[];
+    disallowedHeaders?: string[];
+    overrideFetch?: any;
+    failoverStrategy?: string;
+  }) {
+    if (!apiSecret) {
```

Comment:
just for the reference: I think we have a cetain format for api secret that could be validated.

Reply:
Where do I find that format?

Reply:
Ah that is acutally a really good question. I think it is generated in the API (cc @nijikon )

Reply:
I would add it later , lets have it synced it first with other sdks - without additional features

Reply:
or we should add it to other sdks as well

Reply:
Since the SDK can't work without having some API secret, it's a reasonable requirement. Doesn't castle-ruby require secret?

Reply:
it does, just it doesnt check for proper format of it, just a presence of it.  we can always add extra validations to other sdks :)